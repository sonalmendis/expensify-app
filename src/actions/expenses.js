import uuidv1 from 'uuid/v1';
import database from '../firebase/firebase';

// a component will call an action generator like the ones below
// the action generator then returns an object as shown below
// the component also calls the dispatch method to send the returned object to a reducer
// the reducer then changes the redux store

// FOR ASYNC FUNCTIONS WHERE REDUX COMMUNICATES WITH THE DB (FIREBASE) THE ABOVE IS SLIGHTLY DIFFERENT:
// a component will call an action generator like the ones below
// the action generator then returns A FUNCTION
// the component then dispatched THE FUNCTION
// the function runs
// redux-thunk is used as middleware to allow redux to actually be able to use functions

// ADD EXPENSE ACTION
export const addExpense = expense => ({
	type: 'ADD_EXPENSE',
	expense
});

export const startAddExpense = (expenseData = {}) => {
	// returning a function like below usually wouldn't work without redux-thunk
	return dispatch => {
		// setting up the defaults below:
		const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
		const expense = { description, note, amount, createdAt };
		return database
			.ref('expenses')
			.push(expense)
			.then(ref => {
				dispatch(
					addExpense({
						id: ref.key,
						...expense
					})
				);
			});
	};
};

// REMOVE EXPENSE ACTION
export const removeExpense = ({ id } = {}) => ({
	type: 'REMOVE_EXPENSE',
	id
});

// EDIT EXPENSE ACTION
export const editExpense = ({ id, updates }) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates
});

// SET_EXPENSES
export const setExpenses = expenses => ({
	type: 'SET_EXPENSES',
	expenses
});

export const startSetExpenses = () => {
	return dispatch => {
		return database
			.ref('expenses')
			.once('value')
			.then(snapshot => {
				// remember to use .val() to actually view the snapshot properly
				// console.log(snapshot.val());
				const expenses = [];

				// https://firebase.google.com/docs/reference/js/firebase.database.DataSnapshot
				snapshot.forEach(childSnapshot => {
					expenses.push({
						id: childSnapshot.key, // this is the id that comes from the firebase object
						...childSnapshot.val() // the rest of the object details, description, amount etc.
					});
				});

				console.log(expenses);

				// console.log(dispatch(setExpenses(expenses)));
				// MAKE SURE TO CHECK THAT YOU'RE PASSING IN THE RIGHT FORMAT
				// WHETHER ITS AN ARRAY OR OBJECT
				/*
				dispatch(setExpenses({...expenses})) WONT' WORK
				*/

				console.log(dispatch(setExpenses(expenses)));
			});
	};
};
