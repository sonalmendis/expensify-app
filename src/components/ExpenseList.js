import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = props => (
	<div>
		{props.expenses.length === 0 ? (
			<p>No expenses</p>
		) : (
			props.expenses.map((expense, index) => (
				<ExpenseListItem {...expense} key={expense.id} />
				// Remember to use the spread operator above {...expense} ^ since its the same and a far simpler way of passing props like below:
				// NOTE, if you use the spread operator remember to de-structure it in the expenselistitem file
				// <ExpenseListItem expenseDesc={expense.description} expenseAmount={expense.amount} key={expense.id} />
			))
		)}
	</div>
);

/*
Connect will connect the store to a specific component
You can specifiy which part of the store/state you want passed in
*/

const mapStateToProps = state => {
	return {
		expenses: selectExpenses(state.expenses, state.filters)
	};
};
/**
 * The above just uses the 'selectExpenses' function to actually make use of the filter functions
 * The below is what it looked like before, however, the above is needed to actually make use of the filters
 */
// const mapStateToProps = state => {
// 	return {
// 		expenses: state.expenses,
// 		filters: state.filters
// 	};
// };

export default connect(mapStateToProps)(ExpenseList); // ExpenseList gets passed in to connect, it just looks weird cos that's how their API works
