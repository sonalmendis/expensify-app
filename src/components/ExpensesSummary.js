import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expensesTotals';
import numeral from 'numeral';
export const ExpensesSummary = props => {
	if (props.expenses.length === 0) {
		return (
			<div>
				<p>No expenses to currently show</p>
			</div>
		);
	} else {
		const numberOfExpenses = props.expenses.length;
		const expensesTotalAmount = numeral(props.expensesTotal / 100).format('$0,0.00');
		const expenseWordPlural = numberOfExpenses === 1 ? 'expense' : 'expenses';
		return (
			<div>
				<p>
					You have {numberOfExpenses} {expenseWordPlural} totalling {expensesTotalAmount}
				</p>
			</div>
		);
	}
};

const mapStateToProps = state => {
	console.log(state);
	return {
		expenses: selectExpenses(state.expenses, state.filters),
		expensesTotal: selectExpensesTotal(selectExpenses(state.expenses, state.filters))
	};
};

export default connect(mapStateToProps)(ExpensesSummary);
