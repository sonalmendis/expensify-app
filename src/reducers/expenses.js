const expensesReducerDefaultState = []; // example of setting the default state values as a var beforehand
const expensesReducer = (state = expensesReducerDefaultState, action) => {
	switch (action.type) {
		case 'ADD_EXPENSE':
			return [...state, action.expense];
		// the spread operator [...state] works similar to state.concat(action.expense);
		case 'REMOVE_EXPENSE':
			return state.filter(
				({ id }) => id !== action.id
				// console.log(action.id);
				// console.log(id);
			);
		/* filter above is shorthand for:
                state.filter(({id}) => {
                    return id !== action.id
                })
                ({id}) is just a destructured (expense.id)

        */
		case 'EDIT_EXPENSE':
			return state.map(expense => {
				if (expense.id === action.id) {
					return {
						...expense,
						...action.updates
					};
				} else {
					return expense;
				}
			});
		case 'SET_EXPENSES':
			return action.expenses;
		default:
			return state;
	}
};

export default expensesReducer;
