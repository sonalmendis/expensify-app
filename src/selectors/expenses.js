import moment from 'moment';

// Get visible expense
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
	/*
    The way this function works is by going through the states expenses one by one and then console logging
    them based on the filters from the dispatch. The filters are a way to visibly sort through items BUT
    they don't change the state directly which is why its not used in a reducer.
    */
	return expenses
		.filter(expense => {
			const createdAtMoment = moment(expense.createdAt);
			const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
			// console.log(startDate);
			// console.log(startDateMatch);
			const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
			// console.log(text);

			const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());

			return startDateMatch && endDateMatch && textMatch; // Anything that return false will get filtered out
		})
		.sort((a, b) => {
			//https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort
			if (sortBy === 'date') {
				return a.createdAt < b.createdAt ? 1 : -1;
			} else if (sortBy === 'amount') {
				return a.amount < b.amount ? 1 : -1;
			}
		});
};

export default getVisibleExpenses;
