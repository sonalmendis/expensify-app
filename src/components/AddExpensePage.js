import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
	onSubmit = expense => {
		console.log('updated addexpensepage', expense);
		// props.dispatch(addExpense(expense));
		this.props.startAddExpense(expense);
		this.props.history.push('/'); // to re-direct the page
	};
	render() {
		return (
			<div>
				<h1>Add Expense</h1>
				<ExpenseForm
					/* this prop gets sent to the expense form essentialy so the components can communicate in their 
				relevant contexts e.g in this case to add an expense (not to edit it)
				CHECK THE RELEVNT BOOKMARK ON UDEMY FOR INFO
			*/
					onSubmit={this.onSubmit}
				/>
			</div>
		);
	}
}

const mapDispatchToProps = dispatch => ({
	startAddExpense: expense => dispatch(startAddExpense(expense))
});

// prettier-ignore
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
