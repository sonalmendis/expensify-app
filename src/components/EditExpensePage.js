import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
	onSubmit = updates => {
		const id = this.props.expense.id;
		console.log('updated', updates);
		// console.log(this.props.editExpense({ id, updates }));
		this.props.startEditExpense(id, updates);
		this.props.history.push('/'); // to re-direct the page
	};

	onClick = () => {
		// console.log(dispatch(startRemoveExpense(id)));
		const id = this.props.expense.id;
		this.props.startRemoveExpense(id);
		this.props.history.push('/'); // to re-direct the page
	};

	render() {
		return (
			<div>
				<div className="page-header">
					<div className="content-container">
						<h1 className="page-header__title">Edit Expense</h1>
					</div>
				</div>

				<div className="content-container">
					<ExpenseForm
						className="content-container"
						expense={this.props.expense}
						onSubmit={this.onSubmit}
						buttonCopy={'Save Expense'}
					/>
					<button className="button button--secondary" onClick={this.onClick}>
						Remove Expense
					</button>
				</div>
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch, props) => ({
	startEditExpense: (id, updates) => dispatch(startEditExpense(id, updates)),
	startRemoveExpense: id => dispatch(startRemoveExpense(id))
});

const mapStateToProps = (state, props) => {
	return {
		expense: state.expenses.find(expense => expense.id === props.match.params.id)
	};
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
