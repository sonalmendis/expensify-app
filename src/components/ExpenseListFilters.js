import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
	state = {
		calendarFocused: null
	};
	componentDidMount() {
		// console.log(this.props);
	}
	onDatesChange = ({ startDate, endDate }) => {
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	};
	onFocusChange = calendarFocused => {
		console.log(calendarFocused);
		this.setState(() => ({ calendarFocused }));
	};
	onTextChange = e => {
		console.log(e);
		this.props.setTextFilter({ text: e.target.value });
		// console.log(e.target.value);
		/* the 'dispatch' prop is available here because it gets passed down anyway,
it's not explicitly passed down by us but by the 'connect' component
*/
	};
	onSortChange = e => {
		// console.log(e.target.value);
		e.target.value == 'date' ? this.props.sortByDate() : this.props.sortByAmount();
	};
	render() {
		return (
			<div>
				{/* Text filter */}
				<input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
				{/*
				The 'onChange' attribute above is needed to define the input readability type because we constantly 
				change the value of the input (try removing it and typing inside the input to see the effect)
			*/}
				{/* Filter Type */}
				<select
					value={this.props.filters.sortBy}
					/****
				 * Why do we add the value attribute?
				 * If you don't add it, you end up with that's called an uncontrolled component:
					https://reactjs.org/docs/uncontrolled-components.html
					Basically the DOM becomes the source of truth rather than React which is not what we want.
					This way the default value is controlled by React.
				 */
					onChange={this.onSortChange}
				>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				{/* Date Picker */}
				<DateRangePicker
					startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
					startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
					endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
					endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
					onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
					focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
					onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
					numberOfMonths={1}
					isOutsideRange={() => false}
					showClearDates={true}
				/>
			</div>
		);
	}
}

const mapStatetoProps = state => {
	return {
		filters: state.filters
	};
};

const mapDispatchToProps = dispatch => ({
	setTextFilter: text => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount()),
	setStartDate: startDate => dispatch(setStartDate(startDate)),
	setEndDate: endDate => dispatch(setEndDate(endDate))
});
export default connect(
	mapStatetoProps,
	mapDispatchToProps
)(ExpenseListFilters);
