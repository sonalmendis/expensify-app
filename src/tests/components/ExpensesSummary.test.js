import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import selectExpensesTotal from '../../selectors/expensesTotals';

test('should render expenses summary component correctly without any expenses', () => {
	const wrapper = shallow(<ExpensesSummary expenses={[]} />);
	expect(wrapper).toMatchSnapshot();
});

test('should count single expense and show total', () => {
	const wrapper = shallow(
		<ExpensesSummary expenses={[expenses[0]]} expensesTotal={selectExpensesTotal([expenses[0]])} />
	);
	expect(wrapper).toMatchSnapshot();
});

test('should count multiple expenses and add up correct total', () => {
	const wrapper = shallow(<ExpensesSummary expenses={expenses} expensesTotal={selectExpensesTotal(expenses)} />);
	expect(wrapper).toMatchSnapshot();
});
