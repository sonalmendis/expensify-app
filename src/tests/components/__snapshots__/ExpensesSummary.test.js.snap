// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should count multiple expenses and add up correct total 1`] = `
<div>
  <p>
    You have 
    3
     
    expenses
     totalling 
    $255.00
  </p>
</div>
`;

exports[`should count single expense and show total 1`] = `
<div>
  <p>
    You have 
    1
     
    expense
     totalling 
    $5.00
  </p>
</div>
`;

exports[`should render expenses summary component correctly without any expenses 1`] = `
<div>
  <p>
    No expenses to currently show
  </p>
</div>
`;
